src/xrpld/app/misc/HashRouter.cpp

```diff
@@ -128,4 +142,20 @@ HashRouter::shouldRelay(uint256 const& key)
     return s.releasePeerSet();
 }
 
+void
+HashRouter::forEachPeer(
+    uint256 const& key,
+    std::function<bool(PeerShortID)> callback)
+{
+    std::lock_guard lock(mutex_);
+
+    auto& s = emplace(key).first;
+
+    for (auto const peerID : s.peekPeerSet())
+    {
+        if (!callback(peerID))
```

Comment:
How fast are these callbacks - we are running them over a loop inside critical section. Might need to think of an alternative solution, to avoid hogging the `mutex_`. Not a major issue, because as indicated in first sentence - I do not know if these callbacks are fast or slow (or we don't know ?)

Reply:
The callbacks are very fast, but now I don't remember why I wanted to do this under lock.
PeerImp.cpp line 3528
```
        std::set<HashRouter::PeerShortID> peers;
        app_.getHashRouter().forEachPeer(
            mHash, [&](HashRouter::PeerShortID const& peerID) {
                // callback is called under lock, so finish as fast as
                // possible.
                peers.insert(peerID);
                return true;
            });
```
Maybe I wanted avoid making a copy of the `std::set`? But this is effectively making a copy of the set anyway, so that's dumb. I'd guess I originally wrote it with the loop contents inside the callback, but decided that was too slow, and didn't think before just copying.